src/act/abstract_canister_tree.rs

```diff
@@ -209,6 +214,60 @@ impl AbstractCanisterTree {
         vec![funcs, records, services, tuples, type_aliases, variants].concat()
     }
 
+    fn verify_type_refs_defs_are_unique(&self) -> Result<(), Vec<Error>> {
+        let defined_names = self.candid_types.get_defined_names();
+        let duplicates = find_duplicates(&defined_names);
+
+        match duplicates.is_empty() {
+            true => Ok(()),
+            false => Err(duplicates
+                .into_iter()
+                .map(|type_ref| Error::MultipleTypeDefinitions(type_ref.clone()))
+                .collect()),
+        }
+    }
+
+    fn verify_guard_functions_defs_are_unique(&self) -> Result<(), Vec<Error>> {
+        let defined_names: Vec<_> = self
+            .guard_functions
```

Comment:
Implement `get_defined_names` on guard_functions, query methods, update methods, and anywhere else it makes sense